DO NOT MERGE: kvstore rework - new abstractions with allocator (WIP)

tgraf opened the pull request:
TODO
 - [ ] find way to test TTL expiration of keys
 - [ ] rework & complete docs

aanm commented on pkg/kvstore/allocator.go:
```diff
@@ -0,0 +1,502 @@
+type Allocator struct {
+	// Events is a channel which will receive AllocatorEvent as IDs are
+	// added, modified or removed from the allocator
+	Events AllocatorEventChan
```
Who will consume this channel?

tgraf replied:
The identity code which should trigger a policy recalc when a new identity has been allocated in the cluster.

rlenglet commented on pkg/kvstore/events.go:
```diff
@@ -0,0 +1,99 @@
+type EventChan chan KeyValueEvent
+
+// stopChan is the channel used to indicate stopping of the watcher
+type stopChan chan bool
```
All you do is closing this channel. Therefore, this doesn't need to be of type bool:

    type stopChan chan struct{}

rlenglet commented on pkg/kvstore/events.go:
```diff
@@ -0,0 +1,99 @@
+
+	name      string
+	prefix    string
+	stopWatch stopChan
```
I would restrict the direction of the channel here:

    stopWatch <-chan struct{}

tgraf replied:
```
./events.go:102: invalid operation: close(w.stopWatch) (cannot close receive-only channel)
```

rlenglet commented on pkg/kvstore/events.go:
```diff
@@ -0,0 +1,99 @@
+		name:      name,
+		prefix:    prefix,
+		Events:    make(EventChan, chanSize),
+		stopWatch: make(stopChan, 1),
```
Since all you need to do is close this channel, I would make it unbuffered:

    stopWatch: make(chan struct{}, 0),

rlenglet commented on pkg/kvstore/events.go:
```diff
@@ -0,0 +1,99 @@
+
+// Stop stops a watcher previously created and started with StartWatch
+func (w *Watcher) Stop() {
+	w.stopWatch <- true
```
I don't think that's necessary.

rlenglet commented on pkg/kvstore/events.go:
```diff
@@ -0,0 +1,99 @@
+// changes. Name can be set to anything and is used for logging messages. The
+// Events channel is created with the specified sizes. Upon every change
+// observed, a KeyValueEvent will be sent to the Events channel
+func StartWatch(name, prefix string, chanSize int) *Watcher {
```
This function is only called from allocator.go. Its only purpose is to wrap the singleton Client.
I would instead parameterize the allocator to inject the Client, and have it call KVClient.StartWatch.
Modify the signature of KVClient.StartWatch to match the signature of this function, since KVClient.StartWatch is only called from here.

rlenglet commented on pkg/kvstore/events.go:
```diff
@@ -0,0 +1,99 @@
+	return w
+}
+
+// Stop stops a watcher previously created and started with StartWatch
```
I would add a comment that this method is not idempotent. It panics if called twice.

tgraf replied:
I fixed the code instead to catch this

rlenglet commented on pkg/kvstore/kvstore.go:
```diff
@@ -44,13 +38,94 @@ type KVClient interface {
+	ListPrefix(prefix string) (KeyValuePairs, error)
+
+	// StartWatch starts watching for changes in a prefix
+	StartWatch(w *Watcher)
```
This method is only called from events.go's StartWatch function.
Change the signature of this method to match that of the function. See my comments in events.go.

tgraf replied:
What's the benefit of this? It will require me to duplicate the code that's currently in `events.go` into both etcd and consul. I may misunderstand something there.

rlenglet commented on pkg/kvstore/kvstore.go:
```diff
@@ -44,13 +38,94 @@ type KVClient interface {
+	CreateIfExists(condKey, key string, value []byte, lease bool) error
+
+	// ListPrefix returns a list of keys matching the prefix
+	ListPrefix(prefix string) (KeyValuePairs, error)
```
I don't think you can decouple listing the kv pairs and starting the watcher.
For example, in the case of etcd, the watcher must be started with the revision immediately following the range query's response. Here you give no way for the caller to 1) get the revision of the range query, and 2) pass it to the creation of the watcher.

Therefore, I think those must be combined into a single method so that enough context can be preserved to create the watcher properly. See for instance the K8s client API's ListAndWatch.

tgraf replied:
Yes, we need a `ListAndWatch`. For now, the `ListPrefix()` is used for things like the garbage collector which runs every couple of minutes and removes unused identities.

rlenglet commented on pkg/kvstore/kvstore.go:
```diff
@@ -44,13 +38,94 @@ type KVClient interface {
 	GetWatcher(key string, timeSleep time.Duration) <-chan []policy.NumericIdentity
```
Is this still relevant? How that relate to StartWatcher below?

tgraf replied:
This will be removed once all users are gone.

rlenglet commented on pkg/kvstore/kvstore.go:
```diff
@@ -44,13 +38,94 @@ type KVClient interface {
+	// DeleteLease deletes a lease
+	DeleteLease(interface{}) error
+
+	// Close closes the kvstore client
```
Clarify whether this method is idempotent.

tgraf replied:
I will just remove the code altogether.

rlenglet commented on pkg/kvstore/etcd.go:
```diff
@@ -439,6 +434,47 @@ func (e *EtcdClient) DeleteTree(path string) error {
 
+// StartWatch starts watching for changes in a prefix
+func (e *EtcdClient) StartWatch(w *Watcher) {
+	go func(w *Watcher) {
```
The function parameter is not necessary. This closure can refer to the w parameter.

rlenglet commented on pkg/kvstore/etcd.go:
```diff
@@ -439,6 +434,47 @@ func (e *EtcdClient) DeleteTree(path string) error {
+	go func(w *Watcher) {
+		for {
+		recreateWatcher:
+			etcdWatch := e.cli.Watch(ctx.Background(), w.prefix, client.WithPrefix())
```
You are re-creating the watcher for every watcher response.
This won't work for long.
You must keep the watcher and loop on reading from the channel as long as r.Err() is nil.
etcd is different from ZooKeeper...

rlenglet commented on pkg/kvstore/etcd.go:
```diff
@@ -439,45 +375,45 @@ func (e *EtcdClient) DeleteTree(path string) error {
-				if event.Type != mvccpb.PUT || event.Kv == nil {
-					continue
+		recreateWatcher:
+			etcdWatch := e.cli.Watch(ctx.Background(), w.prefix, client.WithPrefix())
```
You removed WithRev. This won't work. You'll certainly miss revisions.

rlenglet commented on pkg/kvstore/keepalive.go:
```diff
@@ -0,0 +1,68 @@
+
+	// KeepAliveInterval is the internal in which the lease is being
+	// renewed. This must be set to a value lesser than the LeaseTTL
+	KeepAliveInterval = time.Duration(5) * time.Minute
```
No need for explicitly converting to Duration:

    KeepAliveInterval = 5 * time.Minute

Same for all other duration calculations below.

rlenglet commented on pkg/kvstore/keepalive.go:
```diff
@@ -0,0 +1,68 @@
+
+var (
+	// LeaseTTL is the time-to-live ofthe lease
+	LeaseTTL int64 = 60 * 15 // 15 minutes
```
Why not make this a Duration?

rlenglet commented on pkg/kvstore/keepalive.go:
```diff
@@ -0,0 +1,68 @@
+	// LeaseTTL is the time-to-live ofthe lease
+	LeaseTTL int64 = 60 * 15 // 15 minutes
+
+	// KeepAliveInterval is the internal in which the lease is being
```
internal -> interval
lesser -> lower

tgraf replied:
I kept lesser, lower is for height

rlenglet commented on pkg/kvstore/keepalive.go:
```diff
@@ -0,0 +1,68 @@
+				// once we have client and lease, keep alive in regular interval
+				sleepTime = KeepAliveInterval
+				if err := KeepAlive(leaseInstance); err != nil {
+					log.Warningf("Unable to keep lease alive: %s", err)
```
log.WithError(err)...

tgraf replied:
We are not using fields yet so I'd prefer to stay consistent until we switch over.

rlenglet commented on pkg/kvstore/client.go:
```diff
@@ -20,8 +20,14 @@ import (
-var Client KVClient
+var (
+	// Client is the instance of the key-value as configured
+	Client KVClient
```
Don't export this. Instead, define a Client() accessor function, to make sure this is not overwritten accidentally.
I'd use sync.Once.Do in that Client() function, to ensure that this is a singleton.

rlenglet commented on pkg/kvstore/client.go:
```diff
@@ -55,5 +61,25 @@ func initClient() error {
 }
+
+// CloseClient closes the kvstore client
+func CloseClient() {
```
This is only called by test code. Move it into a test package?

rlenglet replied:
Also, this method is exported whereas initClient is not. Why the difference?

tgraf replied:
Removed this code.

rlenglet commented on pkg/kvstore/client.go:
```diff
@@ -55,5 +61,25 @@ func initClient() error {
+
+// CloseClient closes the kvstore client
+func CloseClient() {
+	if Client != nil {
```
This is not threadsafe, if called from another thread than the one that called initClient.

rlenglet commented on pkg/kvstore/consul.go:
```diff
@@ -361,6 +354,59 @@ func (c *ConsulClient) GASNewL3n4AddrID(basePath string, baseID uint32, lAddrID
+			// is different or whether this was a timed out
+			// blocking Get() call
+			if res != nil && (q == nil || q.LastIndex != qo.WaitIndex) {
+				w.Events <- KeyValueEvent{
```
This implementation only produces Modify events. It should also produce Create and Delete events.

rlenglet commented on pkg/kvstore/consul.go:
```diff
@@ -361,6 +354,59 @@ func (c *ConsulClient) GASNewL3n4AddrID(basePath string, baseID uint32, lAddrID
 
+// StartWatch starts watching for changes in a prefix
+func (c *ConsulClient) StartWatch(w *Watcher) {
+	go func(w *Watcher) {
```
No need to pass a Watcher parameter to this closure, it can already access the w parameter to StartWatch.

rlenglet commented on pkg/kvstore/consul.go:
```diff
@@ -361,6 +354,59 @@ func (c *ConsulClient) GASNewL3n4AddrID(basePath string, baseID uint32, lAddrID
+
+		// block Get() calls for 5 seconds maximum
+		qo := consulAPI.QueryOptions{
+			WaitTime: time.Duration(5) * time.Second,
```
WaitTime: 5 * time.Second,

Same for other Duration calculations below.

rlenglet commented on pkg/kvstore/consul.go:
```diff
@@ -361,6 +354,59 @@ func (c *ConsulClient) GASNewL3n4AddrID(basePath string, baseID uint32, lAddrID
+// StartWatch starts watching for changes in a prefix
+func (c *ConsulClient) StartWatch(w *Watcher) {
+	go func(w *Watcher) {
+		nextIndex := uint64(0)
```
var nextIndex uint64

rlenglet commented on pkg/kvstore/consul.go:
```diff
@@ -361,6 +354,59 @@ func (c *ConsulClient) GASNewL3n4AddrID(basePath string, baseID uint32, lAddrID
+			sleepTime := time.Duration(1) * time.Millisecond
+
+			qo.WaitIndex = nextIndex
+			res, q, err := c.KV().Get(w.prefix, &qo)
```
This seems incorrect. I suppose this should be a call to List, not Get, in order to get all the keys with that prefix?!

rlenglet commented on pkg/kvstore/consul.go:
```diff
@@ -407,3 +453,135 @@ func (c *ConsulClient) DeleteTree(path string) error {
+
+// GetPrefix returns the first key which matches the prefix
+func (c *ConsulClient) GetPrefix(prefix string) ([]byte, error) {
+	pairs, _, err := c.KV().List(prefix, nil)
```
You give no option. You may want to consider the RequireConsistent option to true. This will ensure consistent read semantics. This would be consistent with how you query etcd. In your etcd kvstore implementation, you don't specify WithSerializable, so your etcd queries have the default linearizable semantics, equivalent to Consul's RequireConsistent.

rlenglet replied:
Same for all the other Get and List calls.

tgraf replied:
What's the gain? `RequireConsistent` seems to be a lot more expensive.

rlenglet commented on pkg/kvstore/consul.go:
```diff
@@ -407,3 +453,135 @@ func (c *ConsulClient) DeleteTree(path string) error {
+		return nil, err
+	}
+
+	p := KeyValuePairs{}
```
You can allocate a map of the right size from the start:

    p := KeyValuePairs(make(map[string][]byte, len(pairs)))

rlenglet commented on pkg/kvstore/localkeys.go:
```diff
@@ -0,0 +1,82 @@
+	"sync"
+)
+
+type localKey struct {
```
localKey -> refCountValue

tgraf replied:
The structures tracks keys which are in local use so I like the name

rlenglet commented on pkg/kvstore/localkeys.go:
```diff
@@ -0,0 +1,82 @@
+		k.refcnt--
+		if k.refcnt == 0 {
+			delete(lk.keys, key)
+			lk.mutex.Unlock()
```
Replace with:

    defer lk.mutex.Unlock()

Above after the lock.

Same below.

tgraf replied:
defer is slower

rlenglet commented on pkg/kvstore/localkeys.go:
```diff
@@ -0,0 +1,82 @@
+
+// release releases the refcnt of a key. When the last reference was released,
+// the key is deleted
+func (lk *localKeys) release(key string) bool {
```
You may return the interface{} value instead of bool. nil if not found.

tgraf replied:
What is the advantage of that?

rlenglet commented on pkg/kvstore/allocator.go:
```diff
@@ -0,0 +1,539 @@
+// RefillCache invalidates and refills the local cache. Afterwards,
+// Allocator.Cache will contain entries for all IDs to key mappings as found in
+// the KVstore.
+func (a *Allocator) RefillCache() error {
```
Calling this function must be coordinated with stopping and restarting the watcher at the exact revision following the range query. Otherwise, you might either have duplicate updates, or ignore updates.

As I commented on etcd.go, ListPrefix must be combined with the creation of the watcher from the revision following the range query. Therefore, I would refactor RefillCache to stop the watcher and restarting it just after having reset the cache.

No watcher updates must be processed until the cache has been reset. Therefore, I would hold the lock before doing the "list" query, to ensure that the query from etcd and the reset of the cache are done atomically, without any watcher updates being processed in between.

rlenglet commented on pkg/kvstore/allocator.go:
```diff
@@ -0,0 +1,539 @@
+		return err
+	}
+
+	newCache := IDMap{}
```
You can allocate a map of the right size right away:

    newCache := IDMap(make(map[ID]AllocatorKey, len(ids)))

rlenglet commented on pkg/kvstore/allocator.go:
```diff
@@ -0,0 +1,539 @@
+	for k, v := range ids {
+		if id := a.keyToID(k, true); id != 0 {
+			key, err := a.keyType.PutKey(string(v))
+			if err == nil {
```
Log or return any errors?!

rlenglet commented on pkg/kvstore/allocator.go:
```diff
@@ -0,0 +1,539 @@
+	err := CreateOnly(a.idPrefix+strID, k, false)
+	if err == nil {
+		// Add it to local cache
+		a.mutex.Lock()
```
Why aren't the selection of an available ID, the update of the kvstore with that ID, and the update of the local cache done atomically?
It seems like concurrent calls to this method will result in attempting to allocate the same ID (as returned by selectAvailableID) several times, which results in the cache being trashed below every time by any concurrent call.

Every read-update-write block must be protected by a write lock, since the final intent here is to write. This is not the case here. The read part is protected by a read lock, and only the update and write parts are protected by a write lock.
Just acquire the a.mutex.Lock() before calling a.selectAvailableID to eliminate any concurrency issues.

tgraf replied:
Doing the select and update atomically will require to hold a lock throughout the kvstore operations which is something I want to avoid. It will significantly reduce the ability to regenerate endpoints in parallel.

Instead, it seems like we want to make `selectAvailableID()` smart enough to handle concurrent calls by returning a random choice or being aware of the last n IDs handed out which have not been registered yet.

rlenglet commented on pkg/kvstore/allocator.go:
```diff
@@ -0,0 +1,539 @@
+// AllocatorKey is the interface to implement in order for a type to be used as
+// key for the allocator
+type AllocatorKey interface {
+	GetKey() string
```
Please add a comment for every method here. I don't understand how to use those.

rlenglet commented on pkg/kvstore/consul.go:
```diff
@@ -407,43 +344,6 @@ func (c *ConsulClient) StartWatch(w *Watcher) {
-			err error
-		)
-		for {
-			k, q, err = c.KV().Get(key, &qo)
```
Get -> List, to do a recursive query.

rlenglet replied:
Same issue in StartWatch. I can't comment in there as it's not been touched in this PR.

tgraf committed:
kvstore: New kvstore abstraction API

 - Basic lockless operations:
    - Get, GetPrefix, ListPrefix, Set, Delete, DeletePrefix, CreateOnly
    - ListAndWatch, CreateLease, KeepAlive
 - Advanced operations (etcd: lockless, consul: locking required)
    - CreateIfExists
 - Unit tests automatically test etcd and consul backends in single run,
   no need for special make targets anymore
 - Abstracted encoding of keys to allow for binary keys

 - New generic allocator:
   - Maps keys to identifiers using lockless operations in the fast path
   - Uses TTLs to protect reference counting keys in case the node
     disappears
   - Utilizes garbage collector which uses distribtued locks to release
     unused identities
   - Local cache of all identities and keys for fast allocation and
     retrieval
   - Local reference counting if key->id mapping is used multiple times
     from a single node
   - Prepared to support lockless operations with etcd 3.3

Fixes: #915
Fixes: #916

Signed-off-by: Thomas Graf <thomas@cilium.io>

aalemayhu reviewed (commented):
Please ping when this is ready for review.

tgraf committed:
bpf: geneve: Fix convertion of identity ID to hex

For some reason, without the int() cast, the numeric identity is not
properly converted to a hex string which causes issues in the Geneve
TLS validation

Signed-off-by: Thomas Graf <thomas@cilium.io>
